cpp/src/community/approx_weighted_matching_impl.cuh

```diff
@@ -0,0 +1,380 @@
+  cugraph::fill_edge_property(handle, current_graph_view, bool{false}, edge_masks_even);
+  cugraph::edge_property_t<graph_view_t, bool> edge_masks_odd(handle, current_graph_view);
+  cugraph::fill_edge_property(handle, current_graph_view, bool{false}, edge_masks_odd);
+
+  // Mask out self-loop
+  cugraph::transform_e(
+    handle,
+    current_graph_view,
+    cugraph::edge_src_dummy_property_t{}.view(),
+    cugraph::edge_dst_dummy_property_t{}.view(),
+    cugraph::edge_dummy_property_t{}.view(),
+    [] __device__(auto src, auto dst, thrust::nullopt_t, thrust::nullopt_t, thrust::nullopt_t) {
+      return !(src == dst);
+    },
+    edge_masks_even.mutable_view());
+
+  current_graph_view.attach_edge_mask(edge_masks_even.view());
+
+  auto constexpr invalid_partner = invalid_vertex_id<vertex_t>::value;
+  rmm::device_uvector<weight_t> offers_from_partners(
+    current_graph_view.local_vertex_partition_range_size(), handle.get_stream());
+
+  rmm::device_uvector<vertex_t> partners(current_graph_view.local_vertex_partition_range_size(),
+                                         handle.get_stream());
+
+  thrust::fill(handle.get_thrust_policy(), partners.begin(), partners.end(), invalid_partner);
+  thrust::fill(handle.get_thrust_policy(),
+               offers_from_partners.begin(),
+               offers_from_partners.end(),
+               weight_t{0.0});
+
+  rmm::device_uvector<vertex_t> local_vertices(
+    current_graph_view.local_vertex_partition_range_size(), handle.get_stream());
+  detail::sequence_fill(handle.get_stream(),
+                        local_vertices.begin(),
+                        local_vertices.size(),
+                        current_graph_view.local_vertex_partition_range_first());
+
+  using flag_t = uint8_t;
+  edge_src_property_t<graph_view_t, vertex_t> src_key_cache(handle);
+  cugraph::edge_src_property_t<graph_view_t, flag_t> src_match_flags(handle);
+  cugraph::edge_dst_property_t<graph_view_t, flag_t> dst_match_flags(handle);
+
+  if constexpr (graph_view_t::is_multi_gpu) {
+    src_key_cache = edge_src_property_t<graph_view_t, vertex_t>(handle, current_graph_view);
+    src_match_flags =
+      cugraph::edge_src_property_t<graph_view_t, flag_t>(handle, current_graph_view);
+    dst_match_flags =
+      cugraph::edge_dst_property_t<graph_view_t, flag_t>(handle, current_graph_view);
+  }
+
+  vertex_t loop_counter = 0;
+  while (true) {
+    if constexpr (graph_view_t::is_multi_gpu) {
+      update_edge_src_property(handle, current_graph_view, local_vertices.begin(), src_key_cache);
+    }
+
+    //
+    // For each candidate vertex, find the best possible target
+    //
+
+    rmm::device_uvector<vertex_t> candidates(0, handle.get_stream());
+    rmm::device_uvector<weight_t> offers_from_candidates(0, handle.get_stream());
+    rmm::device_uvector<vertex_t> targets(0, handle.get_stream());
+
+    std::forward_as_tuple(candidates, std::tie(offers_from_candidates, targets)) =
```

Comment:
May better add a FIXME statement saying that this can be implemented more efficiently if per_v_transform_reduce_incoming|outgoing_e is updated to support reduction on thrust::tuple.

Reply:
Added.